overlord/snapstate/snapstate.go

```diff
@@ -1200,6 +1200,80 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
 // consider.
 type updateFilter func(*snap.Info, *SnapState) bool
 
+func rearrangeBaseKernelForSingleReboot(kernelTs, bootBaseTs *state.TaskSet) error {
+	var findTaskOfKind = func(ts *state.TaskSet, kind string) (*state.Task, int) {
+		for idx, tsk := range ts.Tasks() {
+			if tsk.Kind() == kind {
+				return tsk, idx
+			}
+		}
+		return nil, -1
+	}
+
+	haveBase, haveKernel := bootBaseTs != nil, kernelTs != nil
+	if !haveBase && !haveKernel {
+		// neither base nor kernel update
+		return nil
+	}
+	if haveBase != haveKernel {
+		// have one but not the other
+		return nil
+	}
+
+	// both kernel and boot base are being updated, reorder link-snap and
+	// auto-connect tasks from both snaps such that we end up with the
+	// following ordering:
+	//
+	// tasks of base and kernel ->
+	//     link-snap(base) ->
+	//        link-snap(kernel)(r) ->
+	//            auto-connect(base) ->
+	//                auto-connect(kernel) ->
+	//                    remaining tasks of base and kernel
+	//
+	// where (r) denotes the task that can effectively request a reboot
+
+	linkSnapKernel, lsKernelIdx := findTaskOfKind(kernelTs, "link-snap")
+	autoConnectKernel, _ := findTaskOfKind(kernelTs, "auto-connect")
+	if linkSnapKernel == nil || autoConnectKernel == nil {
+		return fmt.Errorf("internal error: cannot identify link-snap or auto-connect for the kernel snap")
+	}
+	kernelLanes := linkSnapKernel.Lanes()
+	linkSnapBase, _ := findTaskOfKind(bootBaseTs, "link-snap")
+	autoConnectBase, acBaseIdx := findTaskOfKind(bootBaseTs, "auto-connect")
+	if linkSnapBase == nil || autoConnectBase == nil {
+		return fmt.Errorf("internal error: cannot identify link-snap or auto-connect for the base snap")
+	}
+	baseLanes := linkSnapBase.Lanes()
+
+	for _, lane := range kernelLanes {
+		linkSnapBase.JoinLane(lane)
+		autoConnectBase.JoinLane(lane)
+	}
+	for _, lane := range baseLanes {
+		linkSnapKernel.JoinLane(lane)
+		autoConnectKernel.JoinLane(lane)
+	}
+	// make link-snap base wait for the last pre-link-snap-kernel task
```

Comment:
I went and grepped the source code for a "pre-link-*" task kind, but I found nothing. If we don't have a specific "pre-link" type, I recommend rewriting this comment to something like "_...for the last task before the kernel's "link-snap"_".

It's still not clear why we need to wait for that, though: if the `snap-link(base)` task is the first one we care for (according to the graph from the comment above), it would seem it can be started even earlier.

Reply:
I've tweaked the comment. Ordering is just to ensure that all tasks before link-snap of the kernel have finished before we ru link-snap of the base.